File: steps/4/main/scala/eightyDays.scala/Partner.scala
package eightyDays.scala

abstract class Partner(val name: String)

case class Person(override val name: String, firstName: String) extends Partner(name)

case class LegalEntity(override val name: String, form: String) extends Partner(name)

case class Identification() {
  import java.util.UUID
  val number = UUID.randomUUID()
}

File: steps/7/main/java/eightyDays/java8/Partner.java
package eightyDays.java8;

abstract public class Partner {
    private final String name;

    public Partner(String pName) {
        name = pName;
    }

    public String getName() {
        return name;
    }

    @Override
    public boolean equals(Object pO) {
        if (this == pO) return true;
        if (!(pO instanceof Partner)) return false;

        Partner partner = (Partner) pO;

        return name != null ? name.equals(partner.name) : partner.name == null;
    }

    @Override
    public int hashCode() {
        return name != null ? name.hashCode() : 0;
    }
}
